File: disk-ann/src/main/java/jetbrains/exodus/diskann/DotDistance.java
package jetbrains.exodus.diskann;

import jdk.incubator.vector.FloatVector;
import jdk.incubator.vector.VectorOperators;
import jdk.incubator.vector.VectorSpecies;

final class DotDistance implements DistanceFunction {
    private static final VectorSpecies<Float> species = FloatVector.SPECIES_PREFERRED;

    public double computeDistance(float[] firstVector, float[] secondVector) {
        var sumVector = FloatVector.zero(species);
        var index = 0;

        while (index < species.loopBound(firstVector.length)) {
            var first = FloatVector.fromArray(species, firstVector, index);
            var second = FloatVector.fromArray(species, secondVector, index);
            var mul = first.mul(second);

            sumVector = sumVector.add(mul);
            index += species.length();
        }

        var sum = (double) sumVector.reduceLanes(VectorOperators.ADD);

        while (index < firstVector.length) {
            sum += firstVector[index] * secondVector[index];
            index++;
        }

        return -sum;
    }

    public double computeDistance(
            float[] firstVector,
            int firstVectorFrom,
            float[] secondVector,
            int secondVectorFrom,
            int size
    ) {
        var sumVector = FloatVector.zero(species);
        var index = 0;

        while (index < species.loopBound(size)) {
            var first = FloatVector.fromArray(species, firstVector, index + firstVectorFrom);
            var second = FloatVector.fromArray(species, secondVector, index + secondVectorFrom);
            var mul = first.mul(second);

            sumVector = sumVector.add(mul);
            index += species.length();
        }

        var sum = (double) sumVector.reduceLanes(VectorOperators.ADD);

        while (index < size) {
            sum += firstVector[index + firstVectorFrom] * secondVector[index + secondVectorFrom];
            index++;
        }

        return -sum;
    }


}

File: disk-ann/src/main/kotlin/jetbrains/exodus/diskann/DotDistance.kt
package jetbrains.exodus.diskann

import jdk.incubator.vector.FloatVector
import jdk.incubator.vector.VectorOperators
import jdk.incubator.vector.VectorSpecies

internal class DotDistance : DistanceFunction {
    override fun computeDistance(firstVector: FloatArray, secondVector: FloatArray): Double {
        var sumVector = FloatVector.zero(species)
        var index = 0

        while (index < species.loopBound(firstVector.size)) {
            val first = FloatVector.fromArray(species, firstVector, index)
            val second = FloatVector.fromArray(species, secondVector, index)
            val mul = first.mul(second)

            sumVector = sumVector.add(mul)
            index += species.length()
        }

        var sum = sumVector.reduceLanes(VectorOperators.ADD).toDouble()

        while (index < firstVector.size) {
            sum += firstVector[index] * secondVector[index]
            index++
        }

        return -sum
    }

    override fun computeDistance(
        firstVector: FloatArray,
        firstVectorFrom: Int,
        secondVector: FloatArray,
        secondVectorFrom: Int,
        size: Int
    ): Double {
        var sumVector = FloatVector.zero(species)
        var index = 0

        while (index < species.loopBound(size)) {
            val first = FloatVector.fromArray(species, firstVector, index + firstVectorFrom)
            val second = FloatVector.fromArray(species, secondVector, index + secondVectorFrom)
            val mul = first.mul(second)

            sumVector = sumVector.add(mul)
            index += species.length()
        }

        var sum = sumVector.reduceLanes(VectorOperators.ADD).toDouble()

        while (index < size) {
            sum += firstVector[index + firstVectorFrom] * secondVector[index + secondVectorFrom]
            index++
        }

        return -sum
    }

    companion object {
        private val species: VectorSpecies<Float> = FloatVector.SPECIES_PREFERRED
    }
}